src/Canvas.cc

```diff
@@ -104,8 +104,11 @@ NAN_METHOD(Canvas::New) {
       backend = new ImageBackend(width, height);
   }
   else if (info[0]->IsObject()) {
-    // TODO need to check if this is actually an instance of a Backend to avoid a fault
-    backend = Nan::ObjectWrap::Unwrap<Backend>(info[0]->ToObject());
+    v8::Local<v8::Object> arg = info[0]->ToObject();
+    if(arg.IsEmpty() || arg->InternalFieldCount() == 0){
+      return Nan::ThrowTypeError("Invalid arguments");
+    }
```

Comment:
What you actually want to check here is that the object is an instance of a backend. Something like:

```cpp
(Nan::New(Backend::constructor)->HasInstance(obj))
```

(However I forget how inheritance in backends affects this.)

Reply:
@zbjornson I admit, this was not the perfect solution, but at least it suppresses assertion fault error.

Anyway, I'll add some commits to actually check for the instance, but I'm not sure what behavior should be implemented if a backed instance is passed as the first argument, since `Backend` itself is an abstract class.

My idea now is to implement static method `Nan::Persistent<FunctionTemplate> constructor` in `Backend` class (since it currently doesn't have one) and then in `Backend::New` construct `ImageBackend`, `PdfBackend` or `SvgBackend` based on the argument type. Or you have something other in mind?

> However I forget how inheritance in backends affects this.

I'm not sure either, but current implementation should probably be changed. `Canvas` constructor doesn't check if the passed object is a valid backend, it just wraps it using `nan` module which may result in an error or segfault.

Reply:
Ah right, probably easier to just check each of the implementing classes explicitly. There aren't that many possible Cairo backends, so I'm not worried about this growing out of control.

```cpp
if (Nan::New(ImageBackend::constructor)->HasInstance(info[0]) ||
    Nan::New(PdfBackend::constructor)->HasInstance(info[0]) ||
    Nan::New(SvgBackend::constructor)->HasInstance(info[0])) {
    backend = Nan::ObjectWrap::Unwrap<Backend>(info[0]->ToObject());
}
```

I don't think it's worth the effort of doing it more nicely than that (adding a `Backend::constructor`)...

> Canvas constructor doesn't check if the passed object is a valid backend

Yeah, this slipped through when backend support was added. Thanks for addressing the TODO.